### PR TITLE
Fix request md5 for tagging

### LINKED ISSUE #964: setObjectTagging, removeObjectTagging errors on amazon aws s3
Hello,

Errors with tagging as direct aws s3 client how it works with play.min.io.

Everthing else that i tried also works on aws s3 account.

Thanks :)

**setObjectTagging with tagsMap = {"vuhu":"string"}**
S3Error: The Content-MD5 you specified did not match what we received.
    at Object.parseError (/root/projects/s3/node_modules/minio/dist/main/xml-parsers.js:89:11)
    at /root/projects/s3/node_modules/minio/dist/main/transformers.js:164:22
    at DestroyableTransform._flush (/root/projects/s3/node_modules/minio/dist/main/transformers.js:88:10)
    at DestroyableTransform.prefinish (/root/projects/s3/node_modules/readable-stream/lib/_stream_transform.js:129:10)
    at DestroyableTransform.emit (node:events:394:28)
    at prefinish (/root/projects/s3/node_modules/readable-stream/lib/_stream_writable.js:611:14)
    at finishMaybe (/root/projects/s3/node_modules/readable-stream/lib/_stream_writable.js:620:5)
    at endWritable (/root/projects/s3/node_modules/readable-stream/lib/_stream_writable.js:643:3)
    at DestroyableTransform.Writable.end (/root/projects/s3/node_modules/readable-stream/lib/_stream_writable.js:571:22)
    at IncomingMessage.onend (node:internal/streams/readable:693:10) {
  code: 'BadDigest',
  expecteddigest: 'PHkipudvWos3p1l4VwYI7A==',
  calculateddigest: 'lToqBRvV33vODwQRlq0v0A==',
  requestid: '0HQEMF911SAXB8M5',
  hostid: 'WbX/O5QZus0S1svBbmzW0MzVMNNotqTfQ7mesRwdyx8g62nQKRckDyNHF+xPD06HCOriHMVBKuM=',
  amzRequestid: null,
  amzId2: null,
  amzBucketRegion: null
}

**removeObjectTagging**

S3Error: 204
    at getError (/root/projects/s3/node_modules/minio/dist/main/transformers.js:146:15)
    at /root/projects/s3/node_modules/minio/dist/main/transformers.js:158:14
    at DestroyableTransform._flush (/root/projects/s3/node_modules/minio/dist/main/transformers.js:88:10)
    at DestroyableTransform.prefinish (/root/projects/s3/node_modules/readable-stream/lib/_stream_transform.js:129:10)
    at DestroyableTransform.emit (node:events:394:28)
    at prefinish (/root/projects/s3/node_modules/readable-stream/lib/_stream_writable.js:611:14)
    at finishMaybe (/root/projects/s3/node_modules/readable-stream/lib/_stream_writable.js:620:5)
    at endWritable (/root/projects/s3/node_modules/readable-stream/lib/_stream_writable.js:643:3)
    at DestroyableTransform.Writable.end (/root/projects/s3/node_modules/readable-stream/lib/_stream_writable.js:571:22)
    at IncomingMessage.onend (node:internal/streams/readable:693:10) {
  code: 'UnknownError',
  amzRequestid: null,
  amzId2: null,
  amzBucketRegion: null
}

**getObjectTagging**
Success [ { Key: 'addedfromawsconsole', Value: 'yes it is' } ]

### PR DESCRIPTION
**Pull request by prakashsvmx**
Fixes #964

- Fix broken md5 calculation for tagging requests (S3)

Tested against:

- [x] S3
- [x] MinIO Server ( Local/ play)

### CODE REVIEW

**Comment by ebozduman** [comment]
@prakashsvmx,

I am ready to approve, but it looks like there are some conflicts in your branch.

**Comment by prakashsvmx** [comment]
@ebozduman thank you. I have resolved the conflicts. PTAL.

File: src/main/minio.js [hunk 1]
```diff
@@ -2353,6 +2353,7 @@ export class Client {
     const builder = new xml2js.Builder({ headless:true,renderOpts:{'pretty':false},})
     let payload = builder.buildObject(taggingConfig)
     payload = encoder.encode(payload)
+    headers['Content-MD5'] =toMd5(payload)
```
Comment (harshavardhana) [inline src/main/minio.js:2356]:
```suggestion
    headers['Content-MD5'] = toMd5(payload)
```

File: src/main/helpers.js [hunk 2]
```diff
@@ -387,13 +388,31 @@ export const LEGAL_HOLD_STATUS={
...
+  // browser support is tested only against minio server
+  payLoadBuf = isBrowser ? payLoadBuf.toString() : payLoadBuf 
+  return Crypto.createHash('md5').update(payLoadBuf).digest().toString('base64')
 }
+
 export const toSha256=(payload)=>{
   return Crypto.createHash('sha256').update(payload).digest('hex')
+}
+
+export  const validateStatusCode = (statusCode) =>{
```
Comment (harshavardhana) [inline src/main/helpers.js:407]:
```suggestion
export const validateStatusCode = (statusCode) =>{
```

File: src/main/helpers.js [hunk 3]
```diff
@@ -387,13 +388,31 @@ export const LEGAL_HOLD_STATUS={
...
+  // should be like 200 or [200, 204]
+  if(isArray(statusCode)){
+    const validNumbers = statusCode.filter(isNumber)
+    if(!validNumbers.length){
+      throw new TypeError('statusCode should be an array of http status codes (numbers)')
+    }
+  }else if (!isNumber(statusCode)) {
+    throw new TypeError('statusCode should be a valid number')
+  }
+  return true
```
Comment (harshavardhana) [inline src/main/helpers.js:417]: where is this function used?

File: src/main/helpers.js [hunk 4]
```diff
@@ -387,13 +388,31 @@ export const LEGAL_HOLD_STATUS={
 }
 
 const objectToBuffer = (payload) =>{
-  const payloadBuf = Buffer.from(Buffer.from(JSON.stringify(payload)))
-  return payloadBuf
+  const payloadBuf = Buffer.from(Buffer.from(payload))
+  return  payloadBuf
```
Comment (harshavardhana) [inline src/main/helpers.js:392]:
```suggestion
  return payloadBuf
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: src/main/helpers.js (hunk 3 above)
Comment (prakashsvmx) [reply src/main/helpers.js:417]: Thank you. I will remove it.

**Review by harshavardhana — APPROVED** [review]

**Comment by prakashsvmx** [comment]
@harshavardhana thank you so much for the review.

**Review by ebozduman — APPROVED** [review]
LGTM

**Comment by mapleeit** [comment]
Hi. Could I ask for a new patch for this fix? It really blocks lots of functionalities. Thanks

**Comment by harshavardhana** [comment]
> Hi. Could I ask for a new patch for this fix? It really blocks lots of functionalities. Thanks

Soon!

**Comment by mapleeit** [comment]
Is the `crypto-browserify` missing in `package.json`? After bump up to `7.0.24`, an error has been raised: Cannot find module 'crypto-browserify'

**Comment by harshavardhana** [comment]
> Is the `crypto-browserify` missing in `package.json`? After bump up to `7.0.24`, an error has been raised: Cannot find module 'crypto-browserify'

yep looks like 

> src/main/helpers.js:var Crypto = require('crypto-browserify')

**Comment by harshavardhana** [comment]
@prakashsvmx ^^

**Comment by prakashsvmx** [comment]
@harshavardhana  Thank you so much for the quick fix.

**Pull request merged**